core/tortoise.go

```diff
@@ -0,0 +1,264 @@
+		return false
+	}
+
+	voteFor, voteAgainst = alg.CountVotesInLastLayer(alg.allBlocks[targetBlock]) //??
+
+	/*if err != nil {
+		log.Error("Failed to count votes")
+	}*/
+	if voteFor > alg.LayerVotingAvg() {
+		return true
+	}
+	if voteAgainst > alg.LayerVotingAvg(){
+		return false
+	}
+
+	return originBlock.coin
+}
+
+
+func (alg *Algorithm) getLayerById(layerId uint64) (*Layer, error){
+	if _, ok := alg.layers[layerId]; !ok {
+		return nil, fmt.Errorf("layer id not found %v", layerId)
+	}
+	return alg.layers[layerId], nil
+}
+
+func (alg *Algorithm) CountVotesInLastLayer(block *Block) (uint64, uint64){
+	return block.conVotes, block.proVotes
+	/*
+	var voteFor, voteAgainst uint64 = 0,0
+	l, er := alg.getLayerById(block.layerNum +1)
+	if er != nil {
+		log.Error("%v", er)
+		return 0,0
+	}
+	for i :=0; i < len(l.blocks); i++ {//_,visibleBlock := range l.blocks {
+		if vote, ok := l.blocks[i].blockVotes[block.id]; ok {
+			if vote {
+				voteFor++
+			} else {
+				voteAgainst++
+			}
+		}
+	}
+	return voteFor, voteAgainst*/
+}
+
+
+
+func (alg *Algorithm) createBlockVotingMap(origin *Block) (*bitarray.BitArray, *bitarray.BitArray){
+	//log.Info("creating block voting map for %v", origin.id)
+	blockMap := bitarray.NewBitArray(uint64(alg.totalBlocks))
+	visibilityMap := bitarray.NewBitArray(uint64(alg.totalBlocks))
+	// Count direct voters
+	for blockId,vote := range origin.blockVotes { //todo: check for double votes
+		//todo: assert that block exists
+		targetBlockId := uint64(alg.block2Id[blockId])
+		block := alg.allBlocks[blockId]
+		visibilityMap.SetBit(targetBlockId)
+		targetPosition := alg.visibilityMap[targetBlockId]
+		visibilityMap = visibilityMap.Or(targetPosition.visibility)
+		//log.Info("target map %v new visibility map %v", targetMap.ToNums(), visibilityMap.ToNums())
+		if vote{
+			blockMap.SetBit(targetBlockId)
+			block.proVotes++
+		} else {
+			block.conVotes++
+		}
+	}
+	count := 0
+	ln := len(origin.blockVotes)
+	// Go over all other blocks that exist and calculate the origin blocks votes for them
+	for blockId, targetBlockIdx := range alg.block2Id {
+		if count < ln {
+			if _, ok := origin.blockVotes[blockId]; ok {
+				count++
+				continue
+			}
+		}
+		val, err := visibilityMap.GetBit(uint64(targetBlockIdx))
+		if err != nil{
+			return &blockMap, &visibilityMap //todo: put error
+		}
+		if val {
+			if alg.IsTortoiseValid(origin,blockId ,uint64(targetBlockIdx), visibilityMap) {
+				blockMap.SetBit(uint64(targetBlockIdx))
+			}
+		}
+	}
+	return &blockMap, &visibilityMap
+}
+
+func (alg *Algorithm) countTotalVotesForBlock(targetIdx uint64, visibleBlocks bitarray.BitArray) (uint64, uint64){
+	//targetId := alg.block2Id[target.id]
+	var posVotes, conVotes uint64 = 0,0
+	targetLayer := alg.visibilityMap[targetIdx].layer
+	ln := len(alg.block2Id)
+	for blockIdx:=0; blockIdx< ln; blockIdx++{//_, blockIdx := range alg.block2Id { possible bug what if there is an id > len(alg.block2id)
```

Comment:
remove comment

Reply:
There are few places in the code where you left commented-out code. Please remove them